File: scripts/benchmark_wan21_firstframe_compare.py
#!/usr/bin/env python3
"""Compare first-frame Wan 2.1 CoreML latency before/after."""
import argparse
import re
import subprocess
from pathlib import Path


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser()
    p.add_argument("--before-model", required=True, help="Path to baseline .mlmodelc")
    p.add_argument("--after-model", required=True, help="Path to candidate .mlmodelc")
    p.add_argument("--iters", type=int, default=200)
    p.add_argument("--warmup", type=int, default=20)
    p.add_argument("--compute-units", default="cpu_ne", choices=["cpu_only", "cpu_ne", "cpu_gpu", "all"])
    p.add_argument("--dtype", default="float16", choices=["float16", "float32"])
    p.add_argument("--seed", type=int, default=0)
    p.add_argument("--swift-source", default="scripts/benchmark_wan21_firstframe.swift")
    p.add_argument("--swift-bin", default="/tmp/benchmark_wan21_firstframe")
    return p.parse_args()


def compile_swift(swift_source: str, swift_bin: str) -> None:
    cmd = ["swiftc", "-O", swift_source, "-o", swift_bin]
    print("Compiling benchmark binary:", " ".join(cmd))
    subprocess.run(cmd, check=True)


def run_one(
    model_path: str,
    swift_bin: str,
    iters: int,
    warmup: int,
    compute_units: str,
    dtype: str,
    seed: int,
) -> tuple[float, float, str]:
    cmd = [
        swift_bin,
        model_path,
        "--iters", str(iters),
        "--warmup", str(warmup),
        "--compute-units", compute_units,
        "--dtype", dtype,
        "--seed", str(seed),
    ]
    proc = subprocess.run(cmd, check=True, capture_output=True, text=True)
    out = proc.stdout.strip()
    m = re.search(r"Latency ms: mean=([0-9.]+).*median=([0-9.]+)", out)
    if m is None:
        raise RuntimeError(f"Could not parse benchmark output for {model_path}\n{out}")
    mean = float(m.group(1))
    median = float(m.group(2))
    return mean, median, out


def main() -> None:
    args = parse_args()
    swift_source = Path(args.swift_source)
    swift_bin = Path(args.swift_bin)
    if not swift_source.exists():
        raise SystemExit(f"Swift source not found: {swift_source}")
    compile_swift(str(swift_source), str(swift_bin))

    print("\nRunning BEFORE model benchmark...")
    before_mean, before_median, before_out = run_one(
        model_path=args.before_model,
        swift_bin=str(swift_bin),
        iters=args.iters,
        warmup=args.warmup,
        compute_units=args.compute_units,
        dtype=args.dtype,
        seed=args.seed,
    )
    print(before_out)

    print("\nRunning AFTER model benchmark...")
    after_mean, after_median, after_out = run_one(
        model_path=args.after_model,
        swift_bin=str(swift_bin),
        iters=args.iters,
        warmup=args.warmup,
        compute_units=args.compute_units,
        dtype=args.dtype,
        seed=args.seed,
    )
    print(after_out)

    delta_mean = after_mean - before_mean
    delta_median = after_median - before_median
    pct_mean = (delta_mean / before_mean * 100.0) if before_mean != 0 else 0.0
    pct_median = (delta_median / before_median * 100.0) if before_median != 0 else 0.0

    print("\nComparison Summary")
    print(f"Before mean: {before_mean:.3f} ms")
    print(f"After mean:  {after_mean:.3f} ms")
    print(f"Delta mean:  {delta_mean:+.3f} ms ({pct_mean:+.2f}%)")
    print(f"Before median: {before_median:.3f} ms")
    print(f"After median:  {after_median:.3f} ms")
    print(f"Delta median:  {delta_median:+.3f} ms ({pct_median:+.2f}%)")


if __name__ == "__main__":
    main()

File: scripts/benchmark_wan21_firstframe_compare.swift
import Foundation
import CoreML

enum ArgError: Error, CustomStringConvertible {
    case message(String)

    var description: String {
        switch self {
        case .message(let msg):
            return msg
        }
    }
}

struct Config {
    let beforeModelPath: String
    let afterModelPath: String
    let iterations: Int
    let warmup: Int
    let computeUnits: MLComputeUnits
    let dtypeOverride: MLMultiArrayDataType?
    let seed: UInt64
}

struct LCG {
    private var state: UInt64

    init(seed: UInt64) {
        self.state = seed
    }

    mutating func nextFloat(min: Float, max: Float) -> Float {
        state = 6364136223846793005 &* state &+ 1
        let x = Float((state >> 33) & 0xFFFF_FFFF) / Float(UInt32.max)
        return min + (max - min) * x
    }
}

struct TimingStats {
    let mean: Double
    let median: Double
    let p90: Double
    let p95: Double
    let minV: Double
    let maxV: Double
}

func parseComputeUnits(_ value: String) throws -> MLComputeUnits {
    switch value {
    case "cpu_only":
        return .cpuOnly
    case "cpu_ne":
        return .cpuAndNeuralEngine
    case "cpu_gpu":
        return .cpuAndGPU
    case "all":
        return .all
    default:
        throw ArgError.message("Invalid --compute-units '\(value)'. Use cpu_only|cpu_ne|cpu_gpu|all.")
    }
}

func parseDtype(_ value: String) throws -> MLMultiArrayDataType {
    switch value {
    case "float16":
        return .float16
    case "float32":
        return .float32
    default:
        throw ArgError.message("Invalid --dtype '\(value)'. Use float16|float32.")
    }
}

func parseArgs() throws -> Config {
    let args = CommandLine.arguments
    guard args.count >= 3 else {
        throw ArgError.message(
            """
            Usage:
              benchmark_wan21_firstframe_compare.swift <before.mlmodelc> <after.mlmodelc> [--iters N] [--warmup N] [--compute-units cpu_ne|all|cpu_only|cpu_gpu] [--dtype float16|float32] [--seed N]
            """
        )
    }

    let beforeModelPath = args[1]
    let afterModelPath = args[2]
    var iterations = 200
    var warmup = 20
    var computeUnits: MLComputeUnits = .cpuAndNeuralEngine
    var dtypeOverride: MLMultiArrayDataType?
    var seed: UInt64 = 0

    var i = 3
    while i < args.count {
        let key = args[i]
        switch key {
        case "--iters":
            i += 1
            guard i < args.count, let value = Int(args[i]), value > 0 else {
                throw ArgError.message("Invalid --iters value")
            }
            iterations = value
        case "--warmup":
            i += 1
            guard i < args.count, let value = Int(args[i]), value >= 0 else {
                throw ArgError.message("Invalid --warmup value")
            }
            warmup = value
        case "--compute-units":
            i += 1
            guard i < args.count else { throw ArgError.message("Missing --compute-units value") }
            computeUnits = try parseComputeUnits(args[i])
        case "--dtype":
            i += 1
            guard i < args.count else { throw ArgError.message("Missing --dtype value") }
            dtypeOverride = try parseDtype(args[i])
        case "--seed":
            i += 1
            guard i < args.count, let value = UInt64(args[i]) else {
                throw ArgError.message("Invalid --seed value")
            }
            seed = value
        default:
            throw ArgError.message("Unknown option '\(key)'")
        }
        i += 1
    }

    return Config(
        beforeModelPath: beforeModelPath,
        afterModelPath: afterModelPath,
        iterations: iterations,
        warmup: warmup,
        computeUnits: computeUnits,
        dtypeOverride: dtypeOverride,
        seed: seed
    )
}

func shapeFromConstraint(_ constraint: MLMultiArrayConstraint) -> [Int] {
    return constraint.shape.map { $0.intValue }
}

func product(_ values: [Int]) -> Int {
    return values.reduce(1, *)
}

func fillInput(_ array: MLMultiArray, seed: UInt64) {
    var rng = LCG(seed: seed)
    let count = array.count
    switch array.dataType {
    case .float16:
        let ptr = array.dataPointer.bindMemory(to: Float16.self, capacity: count)
        for i in 0..<count {
            ptr[i] = Float16(rng.nextFloat(min: -1.0, max: 1.0))
        }
    case .float32:
        let ptr = array.dataPointer.bindMemory(to: Float.self, capacity: count)
        for i in 0..<count {
            ptr[i] = rng.nextFloat(min: -1.0, max: 1.0)
        }
    default:
        break
    }
}

func percentile(_ values: [Double], _ p: Double) -> Double {
    precondition(!values.isEmpty)
    let sorted = values.sorted()
    let rank = p * Double(sorted.count - 1)
    let lo = Int(floor(rank))
    let hi = Int(ceil(rank))
    if lo == hi { return sorted[lo] }
    let t = rank - Double(lo)
    return sorted[lo] * (1.0 - t) + sorted[hi] * t
}

func summarize(_ timingsMs: [Double]) -> TimingStats {
    let mean = timingsMs.reduce(0.0, +) / Double(timingsMs.count)
    let median = percentile(timingsMs, 0.5)
    let p90 = percentile(timingsMs, 0.9)
    let p95 = percentile(timingsMs, 0.95)
    let minV = timingsMs.min() ?? 0.0
    let maxV = timingsMs.max() ?? 0.0
    return TimingStats(mean: mean, median: median, p90: p90, p95: p95, minV: minV, maxV: maxV)
}

func benchmarkModel(
    _ model: MLModel,
    provider: MLFeatureProvider,
    warmup: Int,
    iterations: Int
) throws -> TimingStats {
    for _ in 0..<warmup {
        _ = try model.prediction(from: provider)
    }

    var timingsMs: [Double] = []
    timingsMs.reserveCapacity(iterations)
    for _ in 0..<iterations {
        let t0 = DispatchTime.now().uptimeNanoseconds
        _ = try model.prediction(from: provider)
        let t1 = DispatchTime.now().uptimeNanoseconds
        timingsMs.append(Double(t1 - t0) / 1_000_000.0)
    }
    return summarize(timingsMs)
}

func run() throws {
    let cfg = try parseArgs()

    let modelConfig = MLModelConfiguration()
    modelConfig.computeUnits = cfg.computeUnits

    let beforeModel = try MLModel(
        contentsOf: URL(fileURLWithPath: cfg.beforeModelPath),
        configuration: modelConfig
    )
    let afterModel = try MLModel(
        contentsOf: URL(fileURLWithPath: cfg.afterModelPath),
        configuration: modelConfig
    )

    guard let beforeInputDesc = beforeModel.modelDescription.inputDescriptionsByName["latent"],
          let beforeConstraint = beforeInputDesc.multiArrayConstraint else {
        throw ArgError.message("Before model input 'latent' not found or not MLMultiArray")
    }
    guard let afterInputDesc = afterModel.modelDescription.inputDescriptionsByName["latent"],
          let afterConstraint = afterInputDesc.multiArrayConstraint else {
        throw ArgError.message("After model input 'latent' not found or not MLMultiArray")
    }

    let beforeShape = shapeFromConstraint(beforeConstraint)
    let afterShape = shapeFromConstraint(afterConstraint)
    if beforeShape != afterShape {
        throw ArgError.message("Input shapes differ: before=\(beforeShape), after=\(afterShape)")
    }

    let inputType = cfg.dtypeOverride ?? beforeConstraint.dataType
    if inputType != .float16 && inputType != .float32 {
        throw ArgError.message("Only float16/float32 input is supported, got \(inputType.rawValue)")
    }

    let inputArray = try MLMultiArray(
        shape: beforeShape.map(NSNumber.init),
        dataType: inputType
    )
    fillInput(inputArray, seed: cfg.seed)
    let provider = try MLDictionaryFeatureProvider(dictionary: [
        "latent": MLFeatureValue(multiArray: inputArray)
    ])

    let beforeStats = try benchmarkModel(
        beforeModel,
        provider: provider,
        warmup: cfg.warmup,
        iterations: cfg.iterations
    )
    let afterStats = try benchmarkModel(
        afterModel,
        provider: provider,
        warmup: cfg.warmup,
        iterations: cfg.iterations
    )

    let deltaMean = afterStats.mean - beforeStats.mean
    let deltaMedian = afterStats.median - beforeStats.median
    let deltaP95 = afterStats.p95 - beforeStats.p95
    let pctMean = beforeStats.mean != 0 ? (deltaMean / beforeStats.mean * 100.0) : 0.0
    let pctMedian = beforeStats.median != 0 ? (deltaMedian / beforeStats.median * 100.0) : 0.0
    let pctP95 = beforeStats.p95 != 0 ? (deltaP95 / beforeStats.p95 * 100.0) : 0.0

    print("ComputeUnits: \(cfg.computeUnits.rawValue)")
    print("Input shape: \(beforeShape) (\(product(beforeShape)) elems)")
    print("Input dtype: \(inputType == .float16 ? "float16" : "float32")")
    print("Warmup: \(cfg.warmup), Iterations: \(cfg.iterations)")
    print("")
    print("Before model: \(cfg.beforeModelPath)")
    print(String(format: "Latency ms: mean=%.3f median=%.3f p90=%.3f p95=%.3f min=%.3f max=%.3f", beforeStats.mean, beforeStats.median, beforeStats.p90, beforeStats.p95, beforeStats.minV, beforeStats.maxV))
    print("")
    print("After model: \(cfg.afterModelPath)")
    print(String(format: "Latency ms: mean=%.3f median=%.3f p90=%.3f p95=%.3f min=%.3f max=%.3f", afterStats.mean, afterStats.median, afterStats.p90, afterStats.p95, afterStats.minV, afterStats.maxV))
    print("")
    print("Delta (after - before):")
    print(String(format: "mean: %+0.3f ms (%+0.2f%%)", deltaMean, pctMean))
    print(String(format: "median: %+0.3f ms (%+0.2f%%)", deltaMedian, pctMedian))
    print(String(format: "p95: %+0.3f ms (%+0.2f%%)", deltaP95, pctP95))
}

do {
    try run()
} catch {
    fputs("Error: \(error)\n", stderr)
    exit(1)
}
